### PR TITLE
Interpreting ranges for substring 

### PR DESCRIPTION
**Pull request by defstryker**
The changes addresses the points stated in #2493 
I guess this is a more pythonic way of interpreting ranges in slices.

- [x] echo 'good nushell' | str substring '5,12' -> stays the same, returns nushell
- [x] echo 'good nushell' | str substring ',-5' -> changed -5 always means 5th char from the end, so it goes from 0 to length - 5 and returns good nu
- [x] echo 'good nushell' | str substring '5,' -> stays the same, returns nushell
- [x] echo 'good nushell' | str substring ',7' -> stays the same, returns good nu
- [x] echo 'good nushell' | str substring '-5,' -> changed -5 always means 5th char from the end, so it goes from length - 5 to the end and returns shell
- [x] echo 'good nushell' | str substring '-3,-5' -> changed this would try to go from length - 3 to length - 5 and throw the End must be greater than or equal to Start error

### CODE REVIEW

File: crates/nu-cli/src/commands/str_/substring.rs [hunk 1]
```diff
@@ -344,19 +327,20 @@ mod tests {
...
-            expectation("dres", (0, -4)),
-            expectation("res", (0, -3)),
-            expectation("es", (0, -2)),
-            expectation("s", (0, -1)),
+            expectation("", (0, -6)),
+            expectation("a", (0, -5)),
+            expectation("an", (0, -4)),
+            expectation("and", (0, -3)),
+            expectation("andr", (0, -2)),
+            expectation("andre", (0, -1)),
```
Comment (thegedge) [inline crates/nu-cli/src/commands/str_/substring.rs:335]: This makes more sense to me 👍 

Something for us to think about will be how we can do an inclusive end. I'd love to be able to do something like:

```
str substring 5,-1
```

to mean "start at index 5, all the way and including the end". Another option is to interpret a missing value as an inclusive end:

```
str substring 5,
```

Not for this PR, of course 🙂

File: crates/nu-cli/src/commands/str_/substring.rs [hunk 2]
```diff
@@ -141,8 +140,20 @@ fn action(input: &Value, options: &Substring, tag: impl Into<Tag>) -> Result<Val
...
 
-            let start: isize = options.0;
-            let end: isize = options.1;
+            let start: isize = if options.0 < 0 {
+                options.0 + len
+            } else {
+                options.0
+            };
+            let end: isize = if options.1 < 0 {
+                if options.1 < len {
```
Comment (thegedge) [inline crates/nu-cli/src/commands/str_/substring.rs:149]: In this branch, `len` is positive and `options.1` is negative, so this condition will always be true. I also think `0` would be the better default when `options.1 < 0`. What do you think of

```rust
if options.1 >= -len {
    options.1 + len
} else {
    0
}
```

or a little more concise:

```rust
std::cmp::max(len + options.1, 0)
```

?

**Review by thegedge — COMMENTED** [review]
One comment to address, but otherwise I'm 👍 for this behaviour.

@andrasio you added substring, so I want to give you a chance to argue for the old behaviour. I think the new behaviour is far more common in all other languages and systems, at least in my experience.

File: crates/nu-cli/src/commands/str_/substring.rs (hunk 1 above)
Comment (defstryker) [reply crates/nu-cli/src/commands/str_/substring.rs:335]:
```
echo 'good nushell' | str substring 5, -> nushell
```
The above behaviour already exists. We set the default value of a missing `end` value with `isize::max_value()`.

File: crates/nu-cli/src/commands/str_/substring.rs (hunk 2 above)
Comment (defstryker) [reply crates/nu-cli/src/commands/str_/substring.rs]: Nice! I've updated the section with this.

**Comment by jonathandturner** [comment]
I like that we're extending the functionality further here. Awesome.

Let's go ahead and land.

As a follow-up, should we explore extending the range primitive type to be able to describe these kinds of ranges? Then we could use that instead of a string.  eg) `1..10` or `..-10` kinds of ranges

File: crates/nu-cli/src/commands/str_/substring.rs (hunk 1 above)
Comment (thegedge) [reply crates/nu-cli/src/commands/str_/substring.rs:331]: Ahhh, you're totally right. I should've continued past these lines https://github.com/defstryker/nushell/blob/8e62dde93c4bcb3294ac9783b6082d5f3293d9a8/crates/nu-cli/src/commands/str_/substring.rs#L229-L248 😅

**Comment by thegedge** [comment]
> As a follow-up, should we explore extending the range primitive type to be able to describe these kinds of ranges? Then we could use that instead of a string. eg) 1..10 or ..-10 kinds of ranges

Totally. I had that in mind when I saw this implementation. I think using a range parameter would feel far more natural 🙂

**Review by thegedge — APPROVED** [review]
Looks great, @defstryker ! 🙌

**Comment by PunkIslamist** [comment]
Hey all, just wanted to thank @defstryker for implementing this so incredibly quickly! Thanks a lot 🙏 
Also +1 for using ranges here. As this is effectively a range, it would make the API more consistent to use the actual range type here.

**Comment by jonathandturner** [comment]
This should be a step in the right direction: https://github.com/nushell/nushell/pull/2506

It doesn't supporting starting with a negative number (the parser currently sees that as a flag... needs more work there), but other than that you'll now be able to use variables in ranges and have ranges with open ends (like `..5`)

The next step would be to start accepting the ranges and using them like we do these substrings.

I'm excited, this is cool stuff.

**Pull request merged**